policy/testdata/nested_rule5/policy.yaml

```diff
@@ -0,0 +1,30 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: nested_rule5
+rule:
+  match:
+    - condition: x > 0
+      rule:
+        match:
+          - rule:
+              match:
+                - condition: "x > 2"
+                  output: "true"
```

Comment:
Could we add a failure test case for showing incompatible output types for these nested rules involving optionals? For example, changing this output to `optional.of(true)` should make composition fail.

Reply:
Do you want to show the compilation error, composition error, or runtime error? I think if I put `optional.of(true)` that's a compile error since the output types don't match between rules. If I use `dyn(optional.of(true))` I think that works, but it just means the output type is potentially an `optional.of(optional.of(true))`

Reply:
Compilation error post composition due to output type disagreement was what I was thinking. Since we're using optionals to indicate optional steps, it's likely valuable to include a case where the output itself having an optional value doesn't clobber with the optional steps from nested rules.

Reply:
The check that output types agree for all the subrules happens before composition, IIUC.

Reply:
Oh thanks for pointing that out. The type agreement check occurs during composition for Java, so I assumed it must've been the same here. We should be fine